### PR TITLE
Create differential flame graphs for async-profiler or JFR

### PR DESCRIPTION
**Pull request by wolfs**
When using async-profiler or JFR we can generate differential flame graphs for the stack files generated during "regular" profiling. This changes creates the differential flame graphs at the end of the profile runs, if
- Either more than one version or more than one scenario has been used,
- and JFR or async-profiler are used.

### CODE REVIEW

File: src/main/java/org/gradle/profiler/CompositeProfiler.java [hunk 1]
```diff
@@ -122,4 +122,9 @@ public void calculateGradleArgs(List<String> gradleArgs) {
             }
         };
     }
+
+    @Override
+    public boolean isCreatesStacksFiles() {
+        return delegates.stream().anyMatch(Profiler::isCreatesStacksFiles);
```
Comment (blindpirate) [inline src/main/java/org/gradle/profiler/CompositeProfiler.java:128]: What will happen if a composite profiler of profiler A and B, but profiler A supports stacks file, profiler B doesn't support? Will differential tool still work in this case?

**Review by blindpirate — COMMENTED** [review]
Nice work! I left a few comments, it looks great overall.

File: src/main/resources/org/gradle/profiler/flamegraph/flamegraph.pl [hunk 2]
```diff
@@ -107,8 +107,7 @@
 my $nametype = "Function:";     # what are the names in the data?
 my $countname = "samples";      # what are the counts in the data?
 my $colors = "hot";             # color theme
-my $bgcolor1 = "#eeeeee";       # background color gradient start
-my $bgcolor2 = "#eeeeb0";       # background color gradient stop
+my $bgcolors = "";              # background color theme
```
[the remark this answers is not in the document]
Comment (wolfs) [reply src/main/resources/org/gradle/profiler/flamegraph/flamegraph.pl:110]: Please check out the commit comment. I documented there where the files come from.

File: src/main/java/org/gradle/profiler/CompositeProfiler.java (hunk 1 above)
Comment (wolfs) [reply src/main/java/org/gradle/profiler/CompositeProfiler.java:128]: The differential flame graph generation works on the stack files only. So if any of the profiler generates the stacks, then there will be differential flame graphs as well. I don't think using multiple "actual" profilers does make much sense, and they would overwrite each others stack files currently.

Comment (wolfs) [reply src/main/java/org/gradle/profiler/CompositeProfiler.java:128]: Though using something like a chrome trace profiler and async profiler is fine, and you'll get differential flames for the async profiler stacks.

File: src/main/resources/org/gradle/profiler/flamegraph/flamegraph.pl (hunk 2 above)
[the remark this answers is not in the document]
Comment (blindpirate) [reply src/main/resources/org/gradle/profiler/flamegraph/flamegraph.pl:110]: Thanks for the clarification!

File: src/main/java/org/gradle/profiler/CompositeProfiler.java (hunk 1 above)
Comment (blindpirate) [reply src/main/java/org/gradle/profiler/CompositeProfiler.java:112]: Got it.

File: src/test/groovy/org/gradle/profiler/DifferentialFlameGraphIntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        and:
+        assertGraphsGenerated('upToDate')
+        assertDifferentialGraphsGenerated(['upToDate'], [latestSupportedGradleVersion, minimalSupportedGradleVersion])
+
+
+        where:
+        profiler << ["async-profiler", "jfr"]
+    }
+
+    @Requires({ !OperatingSystem.isWindows() })
```
[the remark this answers is not in the document]
Comment (wolfs) [reply src/test/groovy/org/gradle/profiler/DifferentialFlameGraphIntegrationTest.groovy]: Done.

**Pull request merged**